src/number_theory/cyclotomic/gal.lean

```diff
@@ -0,0 +1,209 @@
+/-
+Copyright (c) 2022 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import number_theory.cyclotomic.zeta
+import ring_theory.polynomial.cyclotomic.eval
+import field_theory.polynomial_galois_group
+/-!
+# Galois group of cyclotomic extensions
+
+In this file, we show the relationship between the Galois group of`K(ζₙ)` and `(zmod n)ˣ`;
+it is always a subgroup, and if the `n`th cyclotomic polynomial is irreducible, they are isomorphic.
+
+# Main results
+
+* `is_primitive_root.aut_to_pow`: the monoid hom that takes an automorphism of a ring to the power
+  it sends that specific primitive root, as a member of `(zmod n)ˣ`.
+* `is_cyclotomic_extension.aut_to_pow_injective`: `is_primitive_root.aut_to_pow` is injective
+  in the case that it's considered over a cyclotomic field extension, where `n` does not divide
+  the characteristic of K. As a corollary, $Gal(K(ζₙ)/K)$ is abelian.
+* `is_cyclotomic_extension.aut_equiv_pow`: If, additionally, the `n`th cyclotomic polynomial is
+  irreducible in K, then `aut_to_pow` is a `mul_equiv`.
+* `gal_{X_pow/cyclotomic}_equiv_units_zmod`: Repackage `aut_to_pow` in terms of `polynomial.gal`.
+
+# References
+
+* https://kconrad.math.uconn.edu/blurbs/galoistheory/cyclotomic.pdf
+
+-/
+
+local attribute [instance] pnat.fact_pos
+local attribute [simp] ring_equiv.to_ring_hom_eq_coe
+
+section ring
+
+variables {S : Type*} [comm_ring S] [is_domain S] {μ : S} {n : ℕ+} (hμ : is_primitive_root μ n)
+          (R : Type*) [comm_ring R] [algebra R S]
+
+/-- The `monoid_hom` that takes an automorphism to the power of μ that μ gets mapped to under it. -/
+@[simps {attrs := []}] noncomputable def is_primitive_root.aut_to_pow  :
+  (S ≃ₐ[R] S) →* (zmod n)ˣ :=
```

Comment:
Is it possible to write this as a composition `(S ≃ₐ[R] S) →* roots_of_unity n S →* (zmod n)ˣ`?

Reply:
I guess the components would have to be `(S ≃ₐ[R] S) →* roots_of_unity n S →* zmod n`, but then the composition can be to the units group through the lift. And we can upgrade both of these independently to `≃*`s, and that should come together to the last equiv finally. I'll give this a shot!

Reply:
I'm confused maths-wise about this, actually; `roots_of_unity n S →* zmod n` certainly exists, and it's a map from a group so it should go to actually `(zmod n)ˣ`, but for example `ζ⁰` doesn't map to the unit group! Maybe this will be clear when I write down the Lean code, but I'm not sure what I'm missing

Reply:
ahh, d'oh, this map would take 1 → 0, so it cannot be. maybe this filtering works through `↑(primitive_roots)`? but not sure if that's worth it

Reply:
These two groups are cyclic groups of the same order (let's assume we have a primitive root), so they are isomorphic. In general there is no "canonical isomorphism", but several of them. Writing an homomorphism boils down to choose the image of a generator, and the hom is an iso iff the image of a generator is  generator, so you have `n` morphisms and `\phi n` isomorphisms.

Since we choose `zeta`, and `1` is a nice generator of `(zmod n)`, you can take the iso that sends `zeta` to `1`: it has nothing special, but it is an iso.

Reply:
Yes, but this is * to + and thus inconvenient in Lean; especially as I then have to lift it to the units group, and I'm not sure how automatically that will work.

Reply:
Ah yes, in Lean it's a pain to say that the two groups are isomorphic. But stating that the automorphisms group of `roots_of_unity` is `(zmod n)ˣ` should be easy.